File: adventofcode/2020/16/16.cpp
#include <bits/stdc++.h>
using namespace std;

string join(vector<string> strings, string separator) {
    string joined = strings[0];
    for(int i = 1; i<strings.size(); i++) joined = joined+separator+strings[i];
    return joined;
}

vector<string> split(string value, string separator) {
    vector<string> result; int found; int i = 0;
    while((found = value.find(separator, i)) != string::npos) {
        result.push_back(value.substr(i,found-i));
        i = found+separator.size();
    }
    result.push_back(value.substr(i, value.size()-i));
    return result;
}

bool getLine(string &line) {
    getline(cin, line);
    return line != "";
}

pair<int, int> getRange(string range) {
    vector<string> nums = split(range, "-");
    return make_pair(stoi(nums[0]), stoi(nums[1]));
}

vector<int> getTicket(string line) {
    vector<int> ticket;
    vector<string> values = split(line, ",");
    for(int i = 0; i<values.size(); i++) ticket.push_back(stoi(values[i]));
    return ticket;
}

int main() {
    int sol1 = 0;
    vector<pair<pair<int, int>, pair<int, int> > > ranges;
    vector<string> ranges_names;
    vector<vector<int> > tickets;
    string line;
    while(getLine(line)) {
        vector<string> parts;
        parts = split(line, ":");
        ranges_names.push_back(parts[0]);
        parts = split(parts[1], " ");
        ranges.push_back(make_pair(getRange(parts[1]), getRange(parts[3])));
    }
    getLine(line);
    getLine(line);
    vector<int> your_ticket = getTicket(line);
    getLine(line);
    getLine(line);
    while(getLine(line)){
        vector<int> ticket = getTicket(line);
        bool all_ok = true;
        for(int i = 0; i<ticket.size(); i++) {
            bool found = false;
            for(int j = 0; j<ranges.size(); j++) {
                if ((ranges[j].first.first <= ticket[i] && ticket[i] <= ranges[j].first.second) || 
                (ranges[j].second.first <= ticket[i] && ticket[i] <= ranges[j].second.second)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                sol1 += ticket[i];
                all_ok = false;
            }
        }
        if (all_ok) tickets.push_back(ticket);
    }
    map<int, int> field_index_to_range_index, range_index_to_field_index;
    while (field_index_to_range_index.size() != ranges.size()) {
        for (int field = 0; field < ranges.size(); field++) {
            if (field_index_to_range_index.find(field) != field_index_to_range_index.end()) continue;
            int c = 0, selected_range;
            for (int range = 0; range < ranges.size(); range++) {
                if (range_index_to_field_index.find(range) != range_index_to_field_index.end()) continue;
                bool all_ok = true;
                for(int ticket = 0; ticket < tickets.size(); ticket++) {
                    if (!(ranges[range].first.first <= tickets[ticket][field] && tickets[ticket][field] <= ranges[range].first.second) and 
                    !(ranges[range].second.first <= tickets[ticket][field]  && tickets[ticket][field]  <= ranges[range].second.second)) {
                        all_ok = false;
                        break;
                    }
                }
                if (all_ok) {
                    selected_range = range;
                    c++;
                    if (c > 1) break;
                }
            }
            if(c == 1) {
                field_index_to_range_index[field] = selected_range;
                range_index_to_field_index[selected_range] = field;
                break;
            }
        }
    }
    long long sol2 = 1;
    for(int i = 0; i<ranges_names.size(); i++) {
        vector<string> name_words = split(ranges_names[i], " ");
        if (name_words[0] == "departure") {
            sol2 *= your_ticket[range_index_to_field_index[i]];
        }
    }
    cout << "Problem1: " << sol1 << endl << "Problem2: " << sol2 << endl;
}

File: adventofcode/2020/16/16.py
ranges = []
names = []
line = input().rstrip()
while line:
    names.append(line.split(': ')[0])
    line = line.split(': ')[1].split(' ')
    ranges.append([[int(line[0].split('-')[0]), int(line[0].split('-')[1])], [int(line[2].split('-')[0]), int(line[2].split('-')[1])]])
    line = input().rstrip()
line = input().rstrip()
my_ticket = [int(i) for i in input().rstrip().split(',')]
line = input().rstrip()
line = input().rstrip()
line = input().rstrip()
sol1 = 0
tickets = []
while line:
    line_ok = True
    for i in [int(j) for j in line.rstrip().split(',')]:
        ok = False
        for r in ranges:
            if (r[0][0] <= i and i <= r[0][1]) or (r[1][0] <= i and i <= r[1][1]):
                ok = True
        if not ok:
            sol1 += i
            line_ok = False
    if line_ok:
        tickets.append([int(j) for j in line.rstrip().split(',')])
    line = input()
positions_found = {}
while len(positions_found)<len(ranges):
    for i in range(len(ranges)): # ticket position
        if i not in positions_found:
            possible_sols = []
            for j in range(len(ranges)): # ranges position
                if j not in positions_found.values():
                    all_ok = True
                    for ticket in tickets:
                        if not (ranges[j][0][0] <= ticket[i] <= ranges[j][0][1] or ranges[j][1][0] <= ticket[i] <= ranges[j][1][1]):
                            all_ok = False
                    if all_ok:
                        possible_sols.append(j)
            if len(possible_sols) == 1:
                positions_found[i]=possible_sols[0]
sol2 = 1
for i in positions_found:
    if names[positions_found[i]].split(' ')[0] == 'departure':
        sol2 *= my_ticket[i]
print("Problem1: {}\nProblem2: {}".format(sol1, sol2))